### PR TITLE
Update of morphman to anki version 2.1.35

### PR DESCRIPTION
**Pull request by dmgerman**
- updated settings to use new anki API
- added support to export morphs to sqlite3
- added optional support to ignore narrow parentheses (not available in GUI yet)
- added profiling code (disabled by default)
- print and log the name of the mecab being used

### CODE REVIEW

File: morph/readability.py [hunk 1]
```diff
@@ -282,6 +282,8 @@ def parse_text(text):
                     line_missing_morphs = set()
                     for m in parsed_morphs:
                         # Count morph for word report
+                        log_fp.write('Morph: ' + str(m) + '\n')
```
Comment (ianki) [inline morph/readability.py:285]: Let's drop this change, since it slows down processing.

File: morph/morphemes.py [hunk 2]
```diff
@@ -441,3 +446,105 @@ def analyze2str(self):  # m Str
...
+    conn = connect_db(dbName)
+    with conn:
+        cur = conn.cursor()
+        # fortunately morphman saves "all" first
+        # we only need to save this database, the others
+        # are all subsets of it
+        if (tname == 'all'):
+            save_db_all_morphs(cur, db)
+            tname = 'locations'
+            save_db_locations(cur, db, tname)
```
Comment (ianki) [inline morph/morphemes.py:549]: Are you sure you want to overwrite 'tname'?

Why not just use: save_db_locations(cur, db, 'locations') ?

File: morph/main.py [hunk 3]
```diff
@@ -105,7 +114,8 @@ def mkAllDb(all_db=None):
         N_enabled_notes += 1
 
         mats = [(0.5 if ivl == 0 and ctype == 1 else ivl) for ivl, ctype in
-                db.execute('select ivl, type from cards where nid = :nid', nid=nid)]
+#                db.execute('select ivl, type from cards where nid = :nid', nid=nid)]
```
Comment (ianki) [inline morph/main.py:117]: nit: just delete the commented line

File: morph/main.py [hunk 4]
```diff
@@ -409,8 +419,13 @@ def updateNotes(allDb):
                 (tags_, flds_, sfld, csum, now, mw.col.usn(), nid))
 
     mw.progress.update(label='Updating anki database...')
+    # anki is moving away from this types of queries... so let us redo it in a traditional form
+    tuples = map(lambda x: (x['tags'], x['flds'], x['sfld'], x['csum'],x['now'], x['usn'], x['nid']), ds)
     mw.col.db.executemany(
-        'update notes set tags=?, flds=?, sfld=?, csum=?, mod=?, usn=? where id=?', ds)
+        # replaced this form with the tuples I create to avoid any unexpected error due to order of
+        # attributes in ds changing
+#        'update notes set tags=?, flds=?, sfld=?, csum=?, mod=?, usn=? where id=?', ds)
```
Comment (ianki) [inline morph/main.py]: nit: just delete the line

File: morph/main.py [hunk 5]
```diff
@@ -426,8 +441,12 @@ def updateNotes(allDb):
                 ds.append((due_, now,
                            mw.col.usn(), cid))
 
+    # replace old way to do sql with the standard way
+    tuples = map(lambda x: (x['due'], x['now'], x['usn'], x['cid']), ds)
     mw.col.db.executemany(
-        'update cards set due=?, mod=?, usn=? where id=?', ds)
+#        'update cards set due=?, mod=?, usn=? where id=?', ds)
```
Comment (ianki) [inline morph/main.py]: nit: same here

File: morph/preferences.py [hunk 6]
```diff
@@ -14,11 +14,14 @@ def get_preference(key, model_id=None, deck_id=None):
     except KeyError:
         return _get_anki_json_config(key)
 
-
 def update_preferences(jcfg):
-    original = mw.col.conf['addons']['morphman'].copy()
-    mw.col.conf['addons']['morphman'].update(jcfg)
-    if not mw.col.conf['addons']['morphman'] == original:
+# update with new API
+#    original = mw.col.conf['addons']['morphman'].copy()
```
Comment (ianki) [inline morph/preferences.py:19]: nit: remove commented lines

File: morph/preferences.py [hunk 7]
```diff
@@ -138,11 +141,18 @@ def jcfg_default():
...
-    conf = mw.col.conf['addons']['morphman']
-    assert conf, 'Tried to use jcfgMods before profile loaded'
-    return conf
+# retrieving the configuration using get_config is very expensive operation
+# instead, save it 
+configData = None
 
+def _jsonConfig():
+    # replacel old style with new style
+    #conf = mw.col.conf['addons']['morphman']
```
Comment (ianki) [inline morph/preferences.py:148]: nit: remove commented lines

File: morph/morphemes.py [hunk 8]
```diff
@@ -125,15 +125,16 @@ def find_class(self, cmodule, cname):
...
 square_brackets_regex = re.compile(r'\[[^\]]*\]')
-round_brackets_regex = re.compile(r'（[^）]+）')
+# support for either type of parenthesis.. not sure what would be faster... written like this
+# or split into two regular expressions, or have an or in the regular expression
+# (r'（[^）]+）') # regular expression for wide parenthesis
+round_brackets_regex = re.compile(r'[\(（][^）\)]+[）\)]') 
 
 def getMorphemes(morphemizer, expression, note_tags=None):
     if cfg('Option_IgnoreBracketContents'):
-        if square_brackets_regex.search(expression):
```
Comment (ianki) [inline morph/morphemes.py:129]: I think the 'if' conditions were added intentionally to reduce new string object allocations if they weren't needed.

File: morph/morphemes.py [hunk 9]
```diff
@@ -125,15 +125,16 @@ def find_class(self, cmodule, cname):
         return pickle.Unpickler.find_class(self, cmodule, cname)
 
 square_brackets_regex = re.compile(r'\[[^\]]*\]')
-round_brackets_regex = re.compile(r'（[^）]+）')
+# support for either type of parenthesis.. not sure what would be faster... written like this
+# or split into two regular expressions, or have an or in the regular expression
+# (r'（[^）]+）') # regular expression for wide parenthesis
+round_brackets_regex = re.compile(r'[\(（][^）\)]+[）\)]')
```
Comment (ianki) [inline morph/morphemes.py]: I'd prefer to keep these as separate options.

File: morph/morphemes.py (hunk 8 above)
Comment (ianki) [reply morph/morphemes.py:129]: nvm, I see you made that change in the following patch.

File: morph/config.py [hunk 10]
```diff
@@ -94,6 +94,12 @@
     'new card merged fill': False,
     # k+1 by default. this mostly is to boost performance of 'next new card feature'
     'new card merged fill min due': 100000,
+
+    # round brackets
+    #Option_IgnoreSlimRoundBracketContents : False,
```
Comment (ianki) [inline morph/config.py:99]: Should these lines _be_ commented out?

**Review by ianki — CHANGES_REQUESTED** [review]
I left a few comments, please take a look.

File: morph/morphemes.py (hunk 9 above)
Comment (ianki) [reply morph/morphemes.py]: nvm, I see this was done in the next patch.

**Comment by dmgerman** [comment]
Thanks. I'll clean up the changes. Some issues are because I wrote commits that overrode previous commits. I tried to squash almost everything related. Let me create a new pull request where I address your comments. And clean up the old code.

about the IgnoreBrackets options. I'll profile it and see what is better. I assume speed is more important than memory.

**Comment by joestrong** [comment]
I have done some PRs recently that helps improve some features for at least 2.1.26 too, including the brackets feature, batch play etc. if anyone wants to take a look?

Not sure who the maintainers are here, who can merge etc.

**Comment by ianki** [comment]
> I have done some PRs recently that helps improve some features for at least 2.1.26 too, including the brackets feature, batch play etc. if anyone wants to take a look?
> 
> Not sure who the maintainers are here, who can merge etc.

I can push to the main repo, if you want to send PRs.

**Comment by joestrong** [comment]
> > I have done some PRs recently that helps improve some features for at least 2.1.26 too, including the brackets feature, batch play etc. if anyone wants to take a look?
> > 
> > Not sure who the maintainers are here, who can merge etc.
> 
> I can push to the main repo, if you want to send PRs.

Yeah, the most recent 4 PRs before this one are me! Also this one fixes a recent bug: https://github.com/kaegi/MorphMan/pull/174

**Comment by ianki** [comment]
joestrong has already added round bracket support & tests to https://github.com/kaegi/MorphMan/pull/175.

Should we just merge that?

**Pull request closed without merging**